### PR TITLE
Use rapidjson for materials & sounds

### PR DESCRIPTION
**Pull request by Rackover**

### CODE REVIEW

**Comment by Rackover** [comment]
Yes this PR got out of hand. We're doing techset porting now.

**Comment by Rackover** [comment]
TODO list before merging:
- Export only necessary sounds, not everything
- Need PR on iw4X mod builder for new CSVs (no longer need require & additional includes for game map / fxmap)
- Need to check out the R_Warnings thrown on ported maps (mostly "too much lighting") to see if there's something to do on iw3xport side
- Need to restore techmatching on iw4x side


This PR makes more sense with PR https://github.com/XLabsProject/iw4x-client/pull/427

**Comment by Rackover** [comment]
- IW3-ported map have a small shadowmap bug on the borders of the screen, need to investigate as to why this would happen

Found what it was, shadowmap partition ranges from iw4 are not valid for iw3, i can't edit the shader's CSO so I'm making iw4's partition ranges behave more like in CoD4 for now when a CoD4 map is loaded

File: src/Components/Modules/GSC.cpp [hunk 1]
```diff
@@ -106,16 +107,70 @@ namespace Components
...
 		GSC::PatchReference(data, "maps\\mp\\_createfx", "common_scripts\\_createfx");
 		GSC::PatchReference(data, "maps\\mp\\_utility", "common_scripts\\utility");
 		GSC::DumpSounds(data);
+
 		data = Utils::VA("//_createfx generated. Do not touch!!\n%s", data.data());
 	}
+	
+	void GSC::ConvertToStrictCreateFX(std::string& data)
+	{
+		std::regex regex("(?:ent.*|\\{|main\\(\\)|#.*|\\})", std::regex_constants::icase);
```
Comment (diamante0018) [inline src/Components/Modules/GSC.cpp:128]: static?

File: src/Components/Modules/GSC.cpp [hunk 2]
```diff
@@ -106,16 +107,70 @@ namespace Components
...
+			newData += match[0].str() + "\n";
+			data = match.suffix();
+		}
+
+		data = newData;
+	}
+
+	void GSC::PatchSpecularScale(std::string& data)
+	{
+		std::regex regex("setdvar\\( *\"r_specularcolorscale\", *\"([0-9]+\.*[0-9]*)*\" *\\);"s, std::regex_constants::icase);
```
Comment (diamante0018) [inline src/Components/Modules/GSC.cpp:145]: static?

File: src/Components/Modules/IGameWorld.cpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+			name.AddMember("piecesIndices", pieceIndices, allocator);
+
+			glassNames.PushBack(name, allocator);
+		}
+
+		glassData.AddMember("glassNames", glassNames, allocator);
+
+		//
+		/////////////////////////
```
Comment (diamante0018) [inline src/Components/Modules/IGameWorld.cpp:96]: mhh was there an important comment here?

File: src/Components/Modules/IGameWorld.hpp [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+	public:
+		IGameWorld();
+		~IGameWorld();
+
+		const char* getName() override { return "IGameWorld"; };
+		Game::XAssetType getType() override { return Game::XAssetType::ASSET_TYPE_GAMEWORLD_MP; };
+		void dump(Game::IW3::XAssetHeader header) override { Dump(header.gameWorldMp); };
+
+	private:
+		static void Dump(Game::IW3::GameWorldMp* comWorld);
```
Comment (diamante0018) [inline src/Components/Modules/IGameWorld.hpp:16]: oops

File: src/Components/Modules/IGameWorld.cpp [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+		output.Accept(writer);
+
+		const auto& dump = buff.GetString();
+
+		Utils::WriteFile(Utils::VA("%s/gameworld/%s.iw4x.json", AssetHandler::GetExportPath().data(), basename.data()), dump);
+	}
+
+	IGameWorld::IGameWorld()
+	{
+		Command::Add("dumpGameWorld", [] (Command::Params params)
```
Comment (diamante0018) [inline src/Components/Modules/IGameWorld.cpp:111]: could you fix the command handler and make it pass by pointer or reference the params class?
https://github.com/XLabsProject/iw4x-client/blob/develop/src/Components/Modules/Command.cpp#L162

File: src/Components/Modules/MapDumper.cpp [hunk 6]
```diff
@@ -35,24 +36,43 @@ namespace Components
...
 					else if (Utils::StartsWith(entry->asset.header.sound->aliasName, "c4"))
 					{
+						return;
 					}
-					else
-					{
-						//Components::Logger::Print("%d => %s\n", entry->zoneIndex, entry->asset.header.sound->aliasName);
-						AssetHandler::Dump(Game::XAssetType::ASSET_TYPE_SOUND, entry->asset.header);
+					else if (entry->asset.header.sound->head) {
+						if (entry->asset.header.sound->head->soundFile)
```
Comment (diamante0018) [inline src/Components/Modules/MapDumper.cpp:56]: mhh looks weird

File: src/Utils/Entities.cpp [hunk 7]
```diff
@@ -99,6 +102,73 @@ namespace Utils
...
+		}, false);
+
+
+		if (subModelCount > 0) 
+		{
+			// The last two static models will always be the care packages if we added them ourselves
+			auto countWithoutPackages = subModelCount - 2;
+
+			// All values here taken from mp_rust
+			std::unordered_map<std::string, std::string> airdropPalletBrushModel =
```
Comment (diamante0018) [inline src/Utils/Entities.cpp:127]: This looks messed up by accident

File: src/Utils/Utils.cpp [hunk 8]
```diff
@@ -240,6 +240,7 @@ namespace Utils
 		return buffer;
 	}
 
+
```
Comment (diamante0018) [inline src/Utils/Utils.cpp:243]: Ok

File: src/Utils/Image.hpp [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace Utils
+{
+    namespace Image
+    {
+
+        struct vec3 {
```
Comment (diamante0018) [inline src/Utils/Image.hpp:13]: Class

File: src/Components/Modules/IGfxWorld.cpp [hunk 10]
```diff
@@ -189,6 +192,205 @@ namespace Components
...
+										skips++;
+										continue;
+									}
+
+									auto oldModelIndex = aabbTree->smodelIndexes[smodelIndexIndex];
+									auto newModelIndex = indexesTranslation.at(oldModelIndex);
+
+									if (skips == 0 && oldModelIndex != newModelIndex)
+									{
+										printf("");
```
Comment (Rackover) [inline src/Components/Modules/IGfxWorld.cpp:377]: whoopsie

File: src/Components/Modules/IMaterial.cpp [hunk 11]
```diff
@@ -78,62 +103,252 @@ namespace Components
...
-					constantDef.literal[1] *= 0.165f;
-					constantDef.literal[2] *= 1.4f;
-					constantDef.literal[3] *= 3.2f;
+					// These use the speculars to add some rimlight effects to models
+					// But since speculars are regenerated we end up with cod6 speculars with cod4 materials
+					// and cod6 speculars are a bit too bright for 
+
+					constantDef.literal[0] *= 1.2f; // envMapMin
+					constantDef.literal[1] *= 0.2f;  // envMapMax
+					constantDef.literal[2] *= 1.4f;    // engMapExponent
```
Comment (Rackover) [inline src/Components/Modules/IMaterial.cpp:210]: typo

Comment (Rackover) [inline src/Components/Modules/IMaterial.cpp:210]: These might have to be toned down because the trees on mp_bloc are orange... is that related? It sounds like it could be, but it also could be ground lighting

File: src/Components/Modules/IGfxWorld.cpp [hunk 12]
```diff
@@ -628,10 +884,14 @@ namespace Components
...
-				map.dpvs.smodelInsts[i].lightingOrigin[1] = world->sun.sunFxPosition[1];
-				map.dpvs.smodelInsts[i].lightingOrigin[2] = world->sun.sunFxPosition[2];
+				// this is CORRECT! Lighting origin is the place where the light grid gets sampled, and it must be the object's position!
+				// This is how iw3 does it!
+				// Check out 0x62EFF0 (iw3) 
+				// and 0x524C80 (iw4)
+				// (R_SetStaticModelLighting)
+				map.dpvs.smodelInsts[i].lightingOrigin[0] = map.dpvs.smodelInsts[i].bounds.midPoint[0];
+				map.dpvs.smodelInsts[i].lightingOrigin[1] = map.dpvs.smodelInsts[i].bounds.midPoint[1];
+				map.dpvs.smodelInsts[i].lightingOrigin[2] = map.dpvs.smodelInsts[i].bounds.midPoint[2];
```
Comment (Rackover) [inline src/Components/Modules/IGfxWorld.cpp:894]: I think using DrawInst Placement Origin might be a better call to take info from the pivot. Because bound midpoint could sample lighting from a slightly different place. This is generally good though it's just not... specifically good

File: src/Components/Modules/IMaterial.cpp [hunk 13]
```diff
@@ -243,6 +443,21 @@ namespace Components
 		mat.constantTable  = material->constantTable;
 		mat.stateBitTable  = material->stateBitsTable;
 
+		// Glass
+		// this part is like, totally unnecessary. I thought it fixed an issue on broadcast
+		// but then it clearly didn't. So not sure we should keep it.
+		// Also check via surfacetypebits is big bad
```
Comment (Rackover) [inline src/Components/Modules/IMaterial.cpp:449]: It's fine

File: src/Components/Modules/IMaterialTechniqueSet.cpp [hunk 14]
```diff
@@ -115,131 +466,180 @@ namespace Components
...
+        if (name.contains("_zfeather"))
+        {
+            techset->remappedTechniqueSet = Game::DB_FindXAssetHeader(Game::XAssetType::ASSET_TYPE_TECHNIQUE_SET, std::regex_replace(techset->name, zFeatherRegx, "").data()).techniqueSet;
+            iw4Techset->remappedTechniqueSet = Dump(techset->remappedTechniqueSet);
+        }
+        else if (name.contains("_sm"))
+        {
+            techset->remappedTechniqueSet = Game::DB_FindXAssetHeader(Game::XAssetType::ASSET_TYPE_TECHNIQUE_SET, std::regex_replace(techset->name, smRegx, "_hsm").data()).techniqueSet;
+            iw4Techset->remappedTechniqueSet = Dump(techset->remappedTechniqueSet);
+        }
```
Comment (Rackover) [inline src/Components/Modules/IMaterialTechniqueSet.cpp:595]: Subpar remapping and it also leaves out non-speceular materials... but how to do better?

File: src/Components/Modules/IMaterialTechniqueSet.cpp [hunk 15]
```diff
@@ -115,131 +466,180 @@ namespace Components
...
+            }
         }
 
-        for (int i = 0; i < 48; i++)
+        IMaterialTechniqueSet::SaveConvertedTechset(iw4Techset);
+
+        return iw4Techset;
+    }
+
+    std::string IMaterialTechniqueSet::GetDeclarationName(const Game::IW3::MaterialVertexDeclaration* decl) {
```
Comment (Rackover) [inline src/Components/Modules/IMaterialTechniqueSet.cpp:616]: This function somehow gets it wrong, but it's not extremely important

File: src/Components/Modules/IclipMap_t.cpp [hunk 16]
```diff
@@ -494,6 +513,185 @@ namespace Components
...
+		auto bounds = makeCarePackageBounds();
+		Game::IW3::cbrush_t carePackageBrush;
+		carePackageBrush.numsides = 0;
+		carePackageBrush.baseAdjacentSide = &clipMap->brushEdges[brushEdgeIndex];
+		carePackageBrush.contents = 134420032;
+		carePackageBrush.sides = nullptr;
+		bounds.min(carePackageBrush.mins);
+		bounds.max(carePackageBrush.maxs);
+
+		for (int x = 0; x < 2; ++x)
```
Comment (Rackover) [inline src/Components/Modules/IclipMap_t.cpp:646]: use array size instead?

File: src/Components/Modules/MapDumper.cpp (hunk 6 above)
Comment (Rackover) [reply src/Components/Modules/MapDumper.cpp:64]: none of this should be necessary anymore i think... will need to check

File: src/Utils/Entities.cpp [hunk 17]
```diff
@@ -40,7 +41,10 @@ namespace Utils
 			{
 				std::string model = entity["model"];
 
-				if (!model.empty() && model[0] != '*' && model[0] != '?') // Skip brushmodels
+				if (!model.empty()
+					&& model[0] != '*' && model[0] != '?'  // Skip brushmodels
+					&& model != "com_plasticcase_green_big_us_dirt"s // Skip care package (part of team zones)
```
Comment (Rackover) [inline src/Utils/Entities.cpp:46]: not sure about this!

File: src/Utils/Entities.cpp (hunk 7 above)
Comment (Rackover) [reply src/Utils/Entities.cpp:130]: TAB

File: src/Utils/Image.hpp (hunk 9 above)
Comment (Rackover) [reply src/Utils/Image.hpp:13]: why ?

**Comment by diamante0018** [comment]
this is c++ so I think vec3 should be a class since you added functions to it

File: src/Components/Modules/IGfxWorld.cpp (hunk 12 above)
Comment (Rackover) [reply src/Components/Modules/IGfxWorld.cpp:914]: actually i tested it and, no, this is good !
